internal/server/singleprocess/state/prune.go

```diff
@@ -0,0 +1,85 @@
+package state
+
+import (
+	"sync"
+
+	"github.com/hashicorp/go-memdb"
+)
+
+type pruneOp struct {
+	lock         *sync.Mutex
+	table, index string
+	indexArgs    []interface{}
+	max          int
+	cur          *int
+	check        func(val interface{}) bool
+}
+
+// pruneOld uses the types in op to scan the table indicated and prune old records.
+// The op's check function can allow the process to skip records that shouldn't be
+// pruned regardless of their age.
+func pruneOld(memTxn *memdb.Txn, op pruneOp) (int, error) {
+	op.lock.Lock()
+
+	// Easy enough, just exit if we haven't hit the maximum
+	if *op.cur < op.max {
```

Comment:
We should protect against `op.cur == nil`. One suggestion is maybe just add a `Validate` method to `pruneOp` and use ozzo-validate or something to just do some really basic checking across the board.

Reply:
Added Validate(), good call.